### ISSUE TITLE
[Comments] Small DX improvements

### ISSUE THREAD

**Issue by btrautmann** [CQLabs/dcm-action#152]
Hi! Thanks for your help in https://github.com/CQLabs/dcm-action/issues/150. Opening a new issue here for some requests on improvements to the comments system:

- [ ] **Deleting failure comments upon success:** I noticed that currently if a LOC fails and a comment is made, and then succeeds in a subsequent run, that comment is never deleted, which leaves the PR in a funky state. _Note:_ When I tested this, the "fix" of that LOC made it so that file was no longer modified in the PR, i.e it removed that file from the overall diff. This might be why?
- [ ] **Annotations:** It would be awesome to have inline [annotations](https://github.blog/news-insights/product-news/introducing-check-runs-and-annotations/) on PRs which makes reviewing the full suite of errors easier.
- [ ] **Improvements to comments themselves:** Right now, comments link out to the full report which is nice but it would be nicer to have that information readily available within the comment if possible. This could become _unwieldy_ for large diffs with many errors, and at _that_ point it may make sense to link out to a full report but if there are less than some threshold of issues it would be great to just have that be the comment itself.

Happy to discuss any of these in more details. Thank you for the product and this action!

**Comment by incendial** [comment]
1. It should update the existing comment to "success". Was this not the case?
2. Are your talking about this:
<img width="936" alt="Screenshot 2024-10-24 at 14 54 58" src="https://github.com/user-attachments/assets/cc20ba5c-adcb-4fbf-a875-db7f9098fc63">
It should be working out of the box.
3. We can add more info to the comments, but the large number of issues is the main problem. Also, do you happen to see any good examples from other tool? I don't think the UX of that will be good. Also note that the comments have a character limit.

**Comment by btrautmann** [comment]
> 1. It should update the existing comment to "success". Was this not the case?

Correct, this did not occur on the PR I had tested. In the past I've seen this when the file with a previous failure was removed from the diff, but that may not be the case here... just a thought. Another thought, only _failure_ comments were enabled. Could this have prevented the success message from overriding the previous failure?


> 2. Are your talking about this? It should be working out of the box.

Yes, that's what I'm looking for! This didn't occur on the PR I had tested on (but we did get the comments). Note that I only had the failure comments set to `true`.

> 3. We can add more info to the comments, but the large number of issues is the main problem. Also, do you happen to see any good examples from other tool? I don't think the UX of that will be good. Also note that the comments have a character limit.

Not sure if it's a good comparison, but in the past I've found codecov's comments to be pretty helpful. That said, I don't think this needs to be too complicated... I would imagine that a good percentage of PRs will only have a handful of issues, which I suspect would fall within the comment character limit. Is it a known limit? If so you could just check to see if a "detailed version" of the comment would fit, and otherwise just post a summary with a link to the report. FWIW, it would be awesome if this could all be contained in _one_ comment (that summarizes all packages) instead of one-per-package.

**Comment by incendial** [comment]
> Yes, that's what I'm looking for! This didn't occur on the PR I had tested on (but we did get the comments). Note that I only had the failure comments set to true.

@btrautmann have you figured that out?

> Is it a known limit? 

I think it's something like 65536. So probably 95% unreachable, but showing all issues is still probably undesirable. Maybe a small summary of the first 10-20 is enough.

> FWIW, it would be awesome if this could all be contained in one comment (that summarizes all packages) instead of one-per-package.

Can you please elaborate on that? In which case do you get multiple comments?

**Comment by incendial** [comment]
> Deleting failure comments upon success: I noticed that currently if a LOC fails and a comment is made, and then succeeds in a subsequent run, that comment is never deleted, which leaves the PR in a funky state. Note: When I tested this, the "fix" of that LOC made it so that file was no longer modified in the PR, i.e it removed that file from the overall diff. This might be why?

I double checked this and it should delete comments if `pull_request_comment_on_fail` is set to `true` and there are 0 issues. Although if you are getting multiple comments from multiple packages, that can be the reason.

**Comment by incendial** [comment]
Actually no, even if there are multiple packages, the check should still try to delete a comment.

**Comment by incendial** [comment]
Alright, I've added more logs for both 1 and 2 to see where the problem is. For the number 2 I suspect annotations getting an incorrect relative path.

Can you please check it with the enabled debug logs for actions?
